Tests/Unit/UriGeneratorTest.php

```diff
@@ -73,23 +75,129 @@ public function provideGenerateUri()
                     ),
                 ),
             ),
+            // an exception should be thrown if the token provider is not known
             array(
                 '/this/is/{unknown_token}/life',
                 null,
                 array(),
                 array('InvalidArgumentException', 'Unknown token "unknown_token"'),
             ),
+            // an exception should be thrown if the generated URI is not absolute
             array(
                 'this/is/not/absolute',
                 null,
                 array(),
                 array('InvalidArgumentException', 'Generated non-absolute URI'),
             ),
+            // no tokens need to be specified
             array(
                 '/this/is/has/no/tokens',
                 '/this/is/has/no/tokens',
                 array(),
             ),
+            // nothing should happen if allow_empty is true and the value is not empty
+            array(
+                '/{parent}/title',
+                '/foobar_value/title',
+                array(
+                    'parent' => array(
+                        'name' => 'foobar_provider',
+                        'value' => 'foobar_value',
+                        'options' => array(
+                            'allow_empty' => true,
+                        ),
+                    ),
+                ),
+            ),
+            // the empty token should be collapsed when allow_empty is true
+            array(
+                '/{parent}/title',
+                '/title',
+                array(
+                    'parent' => array(
+                        'name' => 'foobar_provider',
+                        'value' => '',
+                        'options' => array(
+                            'allow_empty' => true,
+                        ),
+                    ),
+                ),
+            ),
+            // if the token value is a single "/" then it should be treated as an empty value and
+            // any trailing slash should be collapsed.
+            array(
+                '{parent}/title',
+                '/title',
+                array(
+                    'parent' => array(
+                        'name' => 'foobar_provider',
+                        'value' => '/',
+                        'options' => array(
+                            'allow_empty' => true,
+                        ),
+                    ),
+                ),
+            ),
+            // if the last segment is empty and allow empty is true, then remove the leading slash
+            array(
+                '/{locale}/{parent}',
+                '/de',
```

Comment:
This new.. if the the last segment is empty and allow empty is true, then remove the leading slash